### PR TITLE
[WebUI] Show a modal for connection error.

### PR DESCRIPTION
**Pull request by kz0817**
With this patch, the modal dialog is shown when a
connection error happens. It also try to reconnect
after a while (currently 30sec. after).

This enables users to notice the error and block
the operation.

***
This is a fixed version of #1842.

### CODE REVIEW

File: client/static/js/hatohol_modal.js [hunk 1]
```diff
@@ -40,8 +40,15 @@
 //
 var HatoholModal = function(params) {
   var self = this;
+  self.owner = false;
 
   self.modalId = params.id;
+  if ($("#" + this.modalId).length) {
```
Comment (cosmo0920) [inline client/static/js/hatohol_modal.js]: It seems that `this` should be `self`?

**Comment by cosmo0920** [comment]
This patch causes HatoholModal browser tests failure like this:

```log
 $(npm bin)/mocha-phantomjs "http://localhost:8000/test/index.html?grep=HatoholModal"


  HatoholModal
    ✓ new
    1) with title
    2) with footer
    3) with body


  1 passing (16ms)
  3 failing

  1) HatoholModal with title:
     Error: expected '' to equal 'test title'
      at http://localhost:8000/test/expect.js:102
      at http://localhost:8000/test/expect.js:219
      at http://localhost:8000/test/expect.js:69
      at http://localhost:8000/test/test_hatohol_modal.js:26
      at callFn (http://localhost:8000/test/mocha.js:4562)
      at http://localhost:8000/test/mocha.js:4555
      at http://localhost:8000/test/mocha.js:4974
      at http://localhost:8000/test/mocha.js:5079
      at next (http://localhost:8000/test/mocha.js:4899)
      at http://localhost:8000/test/mocha.js:4909
      at next (http://localhost:8000/test/mocha.js:4844)
      at http://localhost:8000/test/mocha.js:4871
      at done (http://localhost:8000/test/mocha.js:4518)
      at callFn (http://localhost:8000/test/mocha.js:4573)
      at http://localhost:8000/test/mocha.js:4555
      at next (http://localhost:8000/test/mocha.js:4872)
      at http://localhost:8000/test/mocha.js:4876
      at timeslice (http://localhost:8000/test/mocha.js:6483)

  2) HatoholModal with footer:
     Error: expected '' to equal 'test footer'
      at http://localhost:8000/test/expect.js:102
      at http://localhost:8000/test/expect.js:219
      at http://localhost:8000/test/expect.js:69
      at http://localhost:8000/test/test_hatohol_modal.js:35
      at callFn (http://localhost:8000/test/mocha.js:4562)
      at http://localhost:8000/test/mocha.js:4555
      at http://localhost:8000/test/mocha.js:4974
      at http://localhost:8000/test/mocha.js:5079
      at next (http://localhost:8000/test/mocha.js:4899)
      at http://localhost:8000/test/mocha.js:4909
      at next (http://localhost:8000/test/mocha.js:4844)
      at http://localhost:8000/test/mocha.js:4871
      at done (http://localhost:8000/test/mocha.js:4518)
      at callFn (http://localhost:8000/test/mocha.js:4573)
      at http://localhost:8000/test/mocha.js:4555
      at next (http://localhost:8000/test/mocha.js:4872)
      at http://localhost:8000/test/mocha.js:4876
      at timeslice (http://localhost:8000/test/mocha.js:6483)

  3) HatoholModal with body:
     Error: expected '' to equal 'test body'
      at http://localhost:8000/test/expect.js:102
      at http://localhost:8000/test/expect.js:219
      at http://localhost:8000/test/expect.js:69
      at http://localhost:8000/test/test_hatohol_modal.js:44
      at callFn (http://localhost:8000/test/mocha.js:4562)
      at http://localhost:8000/test/mocha.js:4555
      at http://localhost:8000/test/mocha.js:4974
      at http://localhost:8000/test/mocha.js:5079
      at next (http://localhost:8000/test/mocha.js:4899)
      at http://localhost:8000/test/mocha.js:4909
      at next (http://localhost:8000/test/mocha.js:4844)
      at http://localhost:8000/test/mocha.js:4871
      at done (http://localhost:8000/test/mocha.js:4518)
      at callFn (http://localhost:8000/test/mocha.js:4573)
      at http://localhost:8000/test/mocha.js:4555
      at next (http://localhost:8000/test/mocha.js:4872)
      at http://localhost:8000/test/mocha.js:4876
      at timeslice (http://localhost:8000/test/mocha.js:6483)
```

But test runner is too fast to observe Modal dialog is dismissed.

I think that we might have to apply following patch in HatoholModal browser test:

```diff
diff --git a/client/test/browser/test_hatohol_modal.js b/client/test/browser/tes
t_hatohol_modal.js
index 60acd90..14f2679 100644
--- a/client/test/browser/test_hatohol_modal.js
+++ b/client/test/browser/test_hatohol_modal.js
@@ -1,17 +1,24 @@
 describe('HatoholModal', function() {
   var modal;
+  var modalId;
 
   beforeEach(function() {
     modal = undefined;
   });
 
   afterEach(function() {
-    if (modal)
+    if (modal) {
       modal.close();
+
+      (function removeModal(){
+        $(modalId).remove(); // remove modal at once.
+      })();
+    }
   });
 
   it('new', function() {
     var expectedId = "hatohol-modal";
+    modalId = '#' + expectedId;
     modal = new HatoholModal({id: expectedId});
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
@@ -20,6 +27,7 @@ describe('HatoholModal', function() {
   it('with title', function() {
     var expectedId = "hatohol-modal";
     var title = "test title";
+    modalId = '#' + expectedId;
     modal = new HatoholModal({id: expectedId, title: title});
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
@@ -29,6 +37,7 @@ describe('HatoholModal', function() {
   it('with footer', function() {
     var expectedId = "hatohol-modal";
     var footer = "test footer";
+    modalId = '#' + expectedId;
     modal = new HatoholModal({id: expectedId, footer: footer});
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
@@ -38,6 +47,7 @@ describe('HatoholModal', function() {
   it('with body', function() {
     var expectedId = "hatohol-modal";
     var body = "test body";
+    modalId = '#' + expectedId;
     modal = new HatoholModal({id: expectedId, body: body});
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
```

**Comment by ashie** [comment]
> I think that we might have to apply following patch in HatoholModal browser test:

Because modal.close() does modal.remove(), I prefer the following patch:

```js
diff --git a/client/test/browser/test_hatohol_modal.js b/client/test/browser/test_hatohol_modal.js
index 60acd90..cbf6cb9 100644
--- a/client/test/browser/test_hatohol_modal.js
+++ b/client/test/browser/test_hatohol_modal.js
@@ -5,14 +5,15 @@ describe('HatoholModal', function() {
     modal = undefined;
   });
 
-  afterEach(function() {
+  afterEach(function(done) {
     if (modal)
-      modal.close();
+      modal.close(done);
   });
 
   it('new', function() {
     var expectedId = "hatohol-modal";
     modal = new HatoholModal({id: expectedId});
+    modal.show();
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
   });
@@ -21,6 +22,7 @@ describe('HatoholModal', function() {
     var expectedId = "hatohol-modal";
     var title = "test title";
     modal = new HatoholModal({id: expectedId, title: title});
+    modal.show();
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
     expect($('.modal-title').text()).to.be(title);
@@ -30,6 +32,7 @@ describe('HatoholModal', function() {
     var expectedId = "hatohol-modal";
     var footer = "test footer";
     modal = new HatoholModal({id: expectedId, footer: footer});
+    modal.show();
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
     expect($('.modal-footer').text()).to.be(footer);
@@ -39,6 +42,7 @@ describe('HatoholModal', function() {
     var expectedId = "hatohol-modal";
     var body = "test body";
     modal = new HatoholModal({id: expectedId, body: body});
+    modal.show();
     expect(modal).not.to.be(undefined);
     expect($('#' + expectedId)).to.have.length(1);
     expect($('.modal-body').text()).to.be(body);
```

**Comment by ashie** [comment]
@cosmo0920, please commit your preferred patch.
I'll merge them.

**Comment by cosmo0920** [comment]
I prefer to use @ashie's patch.
I've committed it.

**Pull request merged**